container/sagemaker/python_service.py

```diff
@@ -0,0 +1,151 @@
+
+import falcon
+import requests
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+DEFAULT_CONTENT_TYPE = 'application/json'
+DEFAULT_ACCEPT_HEADER = 'application/json'
+CUSTOM_ATTRIBUTES = 'X-Amzn-SageMaker-Custom-Attributes'
+
+Context = namedtuple('Context',
+                     'model_name, model_version, method, rest_uri, grpc_uri, '
+                     'custom_attributes, request_content_type, accept_header')
+
+
+class InvocationResource(object):
+
+    def __init__(self):
+        if 'SAGEMAKER_SAFE_PORT_RANGE' in os.environ:
+            port_range = os.environ['SAGEMAKER_SAFE_PORT_RANGE']
+            parts = port_range.split('-')
+            low = int(parts[0])
+            high = int(parts[1])
+            if low + 2 > high:
+                raise ValueError('not enough ports available in SAGEMAKER_SAFE_PORT_RANGE ({})'
+                                 .format(port_range))
+            self._tfs_grpc_port = str(low)
+            self._tfs_rest_port = str(low + 1)
+        else:
+            self._tfs_grpc_port = '9000'
+            self._tfs_rest_port = '8501'
+
+        self.handler = None
+        self.input_handler = None
+        self.output_handler = None
+        self._import_handlers()
+
+    def on_post(self, req, res):
+        context, data = self._parse_request(req)
+
+        if self.handler:
+            res.status = falcon.HTTP_200
+            res.body = self.handler(data, context)
+        elif self.input_handler and self.output_handler:
+            processed_input = self.input_handler(data, context)
+            response = requests.post(context.rest_uri, data=json.dumps(processed_input))
+            prediction, content_type = self.output_handler(response, context)
+            prediction = json.dumps(prediction)
+
+            res.status = falcon.HTTP_200
+            res.body, res.content_type = prediction, content_type
+
+    def _import_handlers(self):
+        spec = importlib.util.spec_from_file_location('inference', '/opt/ml/model/inference.py')
+        inference = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(inference)
+
+        try:
+            self.handler = inference.handler
+            log.warning('Using handler, input_handler and output_handler will be ignored.')
+        except AttributeError:
+            try:
+                self.input_handler = inference.input_handler
+                self.output_handler = inference.output_handler
+            except AttributeError:
+                log.error('Handlers are not implemented correctly.')
+
+    def _parse_request(self, req):
+        content_type = req.get_header('Content-Type') or DEFAULT_CONTENT_TYPE
+        accept = req.get_header('Accept') or DEFAULT_ACCEPT_HEADER
+        attributes = self._parse_custom_attributes(req.get_header(CUSTOM_ATTRIBUTES))
+
+        context = self._make_context(attributes, content_type, accept)
+        data = req.stream.read().decode('utf-8')
```

Comment:
several issues here...

1. data won't always be text data, so decoding as utf-8 will fail. just passing result of req.stream.read() would be better
2. read() will load everything into memory. for large requests (e.g. in batch), this will cause higher memory usage. maybe we should just pass req.stream and leave it to user to call read (and document that data is "file-like object". if we do this, we should add content_length to the context object as well.